### PR TITLE
Hw 2

### PR DESCRIPTION
**Pull request by Ivan-mech99**
Works somehow

### CODE REVIEW

File: src/network/mt_blocking/ServerImpl.cpp [hunk 1]
```diff
@@ -35,6 +35,7 @@ ServerImpl::~ServerImpl() {}
 
 // See Server.h
 void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
+    num_thr = 4;
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: n_workers

File: src/network/mt_blocking/ServerImpl.cpp [hunk 2]
```diff
@@ -71,22 +72,121 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
         close(_server_socket);
         throw std::runtime_error("Socket listen() failed");
     }
-
+    std::unique_lock<std::mutex> start_lock(socket_mutex);
+    running.store(true);
+    sockets.insert(_server_socket);
+    start_lock.unlock();
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Не нужно

File: src/network/mt_blocking/ServerImpl.cpp [hunk 3]
```diff
@@ -71,22 +72,121 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
     running.store(true);
     _thread = std::thread(&ServerImpl::OnRun, this);
+    _thread.detach();
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Во время Join нужно гарантировать завершение серверного потока => не detach(), join

File: src/network/mt_blocking/ServerImpl.cpp [hunk 4]
```diff
@@ -71,22 +72,121 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
-    _thread.join();
-    close(_server_socket);
+    std::unique_lock<std::mutex> join_lock(socket_mutex);
+    while(running||(!sockets.empty())){
+        _stop.wait(join_lock);
+    }
+}
+
+void ServerImpl::onWork(int client_socket) {
+    std::size_t arg_remains;
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: = 0;

File: src/network/mt_blocking/ServerImpl.cpp [hunk 5]
```diff
@@ -71,22 +72,121 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
+    Protocol::Parser parser;
+    std::string argument_for_command;
+    std::unique_ptr<Execute::Command> command_to_execute;
+        try {
+            int readed_bytes = -1;
+            char client_buffer[4096];
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: = "";

File: src/network/mt_blocking/ServerImpl.cpp [hunk 6]
```diff
@@ -134,15 +234,28 @@ void ServerImpl::OnRun() {
...
-    // Cleanup on exit...
+     //Cleanup on exit...
+     std::unique_lock<std::mutex> onrun_lock1(socket_mutex);
+     sockets.erase(_server_socket);
+     close(_server_socket);
+     //_thread.detach();
+    if(!running&&sockets.empty()){
+         _stop.notify_all();
+    }
+    onrun_lock1.unlock();
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Не нужен

File: src/network/mt_blocking/ServerImpl.cpp [hunk 7]
```diff
@@ -71,22 +72,121 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
+                            _logger->debug("Found new command: {} in {} bytes", parser.Name(), parsed);
+                            command_to_execute = parser.Build(arg_remains);
+                            if (arg_remains > 0) {
+                                arg_remains += 2;
+                            }
+                        }
+
+                        // Parsed might fails to consume any bytes from input stream. In real life that could happens,
+                        // for example, because we are working with UTF-16 chars and only 1 byte left in stream
+                        if (parsed == 0) {
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp:133]: В этом случае нужно начинать читать со смещения readed_bytes

**Review by Fen99 — CHANGES_REQUESTED** [review]

**Review by Fen99 — CHANGES_REQUESTED** [review]
Исправить Put, добавить тест (put с уменьшением размера)

File: src/network/mt_blocking/ServerImpl.cpp [hunk 8]
```diff
@@ -134,15 +235,27 @@ void ServerImpl::OnRun() {
 
         // TODO: Start new thread and process data from/to connection
         {
-            static const std::string msg = "TODO: start new thread and process memcached protocol instead";
-            if (send(client_socket, msg.data(), msg.size(), 0) <= 0) {
-                _logger->error("Failed to write response to client: {}", strerror(errno));
-            }
-            close(client_socket);
+         std::unique_lock<std::mutex> onrun_lock(socket_mutex);
+         if(sockets.size()<num_thr){
```
Comment (Fen99) [inline src/network/mt_blocking/ServerImpl.cpp]: Здесь нужно проверять, что && running, иначе можно добавить нового клиента на завершенном сервере

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: test/storage/StorageTest.cpp [hunk 9]
```diff
@@ -56,6 +56,20 @@ TEST(StorageTest, PutIfAbsent) {
...
+
+    EXPECT_TRUE(storage.Put("KEY11", "val11"));
+
+    EXPECT_TRUE(storage.Put("KEY22", "val22"));
+
+    EXPECT_TRUE(storage.Put("KEY11", "val4"));
+    std::string value;
+    EXPECT_TRUE(storage.Get("KEY11", value)); 
+    EXPECT_TRUE(value == "val4");
+}
```
Comment (Fen99) [inline test/storage/StorageTest.cpp:70]: Еще можно добавить проверку, что KEY22 остался

**Review by Fen99 — APPROVED** [review]
